### ISSUE TITLE
Implement phasic policy gradient (PPG) in alf

### ISSUE THREAD

**Issue by breakds** [HorizonRobotics/alf#965]
# Motivation

PPG is claimed to be significantly better in terms of sample efficiency compared to PPO, and we heard that it has become the default algorithm in OpenAI. Therefore, it would be great if we can have an implementation of PPG in alf so that it can be used as baseline, or combined with other techniques/algorithms to create new solutions for future experiments.

# Tasks

- [x] Study and understand in detail the architecture of PPG
- [x] Implement PPG based on its official (open source) [implementation](https://github.com/openai/phasic-policy-gradient).
- [x] Tune it on cart pole
- [ ] Tune it on procgen, which is the benchmark released by OpenAI, where the original implementation is tested against

**Comment by breakds** [comment]
A few design choices.

If the main network and the auxiliary network are completely separate, the graphic card memory consumption will almost double. Per the paper and OpenAI's official implementation, a compromise (at the cost slightly worse performance) is proposed to **share the encoding network, but detach the gradient update for the auxiliary head**. We can put an option in `PPGAlgorithm` so that the user can choose.

**Comment by breakds** [comment]
![train_iter_off_policy](https://user-images.githubusercontent.com/1111035/128581804-b4496f29-b944-4302-84fb-fef678d6de93.png)

PPG requires running **two different kinds of behavior in the `_update()` loop** above. As discussed with @emailweixu  and @hnyu, it is probably better to override the `train_step` and `calc_loss` of `PPGAlgorithm` to inject the "loop ID" in `train_info`, so that we can decide what kind of behavior we want based on the loop ID.

Also it seems that `preprocess_experience()` looks like the right place to inject such loop ID and set it to `0`.

**Comment by breakds** [comment]
Went through [OpenAI's Implementation](https://github.com/openai/phasic-policy-gradient) in more detail, I created the following diagram to show the network architecture and loss function construction. 

There are always **3 heads** in the network architecture and 2 different loss functions. One of the loss functions is for normal training iteration where roll-outs are performed, and the other one is for aux phase updates (I call it extra iteration so that the "aux" does not confuse with the aux network).

![ppg_dual_network](https://user-images.githubusercontent.com/1111035/129265833-19c85a99-8fed-498d-9d3f-591074ff0e7e.png)

**Comment by breakds** [comment]
As the above is for the "dual" network architecture, there is also a "detached" network. The only difference is that in "detached" network, all 3 heads share the same encoder network, with "aux value head" having the encoder network's gradient detached.

![ppg_detached_network](https://user-images.githubusercontent.com/1111035/129266029-73eb2b8a-007d-43cd-8d43-90259efb9d7a.png)

**Comment by hnyu** [comment]
> Went through [OpenAI's Implementation](https://github.com/openai/phasic-policy-gradient) in more detail, I created the following diagram to show the network architecture and loss function construction.
> 
> There are always **3 heads** in the network architecture and 2 different loss functions. One of the loss functions is for normal training iteration where roll-outs are performed, and the other one is for aux phase updates (I call it extra iteration so that the "aux" does not confuse with the aux network).
> 
> ![ppg_dual_network](https://user-images.githubusercontent.com/1111035/129265833-19c85a99-8fed-498d-9d3f-591074ff0e7e.png)

Why there is a KD divergence loss (on the right) in the normal iteration loss? Normally PPO doesn't have this.

**Comment by breakds** [comment]
> > Went through [OpenAI's Implementation](https://github.com/openai/phasic-policy-gradient) in more detail, I created the following diagram to show the network architecture and loss function construction.
> > There are always **3 heads** in the network architecture and 2 different loss functions. One of the loss functions is for normal training iteration where roll-outs are performed, and the other one is for aux phase updates (I call it extra iteration so that the "aux" does not confuse with the aux network).
> > ![ppg_dual_network](https://user-images.githubusercontent.com/1111035/129265833-19c85a99-8fed-498d-9d3f-591074ff0e7e.png)
> 
> Why there is a KD divergence loss (on the right) in the normal iteration loss? Normally PPO doesn't have this.

And our implementation does not have this term too. Your comment reminds to check the hyper parameters of the OpenAI implementation. Apparently they do have this term in the normal iteration loss, but the hyper parameter coefficient for that KL divergence term is **set to 0**.

```python
def train_fn(env_name="coinrun",
    ....
    arch="dual",  # 'shared', 'detach', or 'dual'
    clip_param=.2,
    kl_penalty=0.0,
    ...)
```

Thanks for pointing this out!

My understanding is that clipping pg term already sort of absorbing the updated policy towards the old (pre-update) policy, and therefore the KL divergence which has the same intention isn't exactly required.

**Comment by breakds** [comment]
Progress so far:

1. Implemented a simplified version that can do training (but converges rather slowly if at all). The simplified version does auxiliary update in every training iteration. I plan to debug the slow convergence problem once the true version is implemented.
2. The major blocker for implementing the true version is that we want to achieve:
  * In normal iteration, train with everything from rollout of **the current iteration**, while pushing the rollout trajectories into the replay buffer.
  * Every 32 iterations (or whatever is configured), the auxiliary update will be executed for everything that is in the replay buffer
  * After each auxiliary update, the replay buffer will be cleared.

To achieve No.2, I think it might be possible to add a method to the Replay Buffer which supports `play_current_iteration` alongside the `play_all` and `sample`.

**Comment by hnyu** [comment]
> As the above is for the "dual" network architecture, there is also a "detached" network. The only difference is that in "detached" network, all 3 heads share the same encoder network, with "aux value head" having the encoder network's gradient detached.
> 
> ![ppg_detached_network](https://user-images.githubusercontent.com/1111035/129266029-73eb2b8a-007d-43cd-8d43-90259efb9d7a.png)

The figure shows that the true value head instead of the aux value head is detached.

**Comment by breakds** [comment]
> > As the above is for the "dual" network architecture, there is also a "detached" network. The only difference is that in "detached" network, all 3 heads share the same encoder network, with "aux value head" having the encoder network's gradient detached.
> > ![ppg_detached_network](https://user-images.githubusercontent.com/1111035/129266029-73eb2b8a-007d-43cd-8d43-90259efb9d7a.png)
> 
> The figure shows that the true value head instead of the aux value head is detached.

Sorry I just notice this comments. Yes that is intended.

There are two components, the policy component and the (true) value component.

1. The policy component has two heads, the policy head and the aux value head.
2. The (true) value component has only one head, the true value head.

In `dual` architecture, each component has their own encoder, while in `shared` architecture, the value component shares the detached encoder. Therefore the graph above.

**Comment by breakds** [comment]
Some updates on my recent effort of establishing PPO baseline on procgen. The biggest problem was that the efficacy quickly reaches "somewhat good" during training but dropped to almost almost-zero-reward at some point really quickly. At that moment the algorithm converges to trivial policy (shooting without moving) - see the video below:

https://user-images.githubusercontent.com/1111035/132711546-ed9125fa-fe81-4f0a-aabd-392692b960d6.mp4

**Comment by breakds** [comment]
Yesterday after adding advantage normalization and replacing huber with MSE, such problem is almost solved until at the end it still converges to the trivial policy. The below graph compares the before (orange) and after (red) training curve. I think it is much alleviated but it seems that the algorithm cannot help converging to trivial policy at the end. 

![average_return_000](https://user-images.githubusercontent.com/1111035/132712433-127c121c-e185-4316-90d1-b7b1f0d160b1.png)

Note that with my current setting, iteration = 4000 is sort of equivalent to the number of total steps in OpenAI's implementation. The new training actually survived until iteration 4000 but "dip" at around 6500. This raises a hypothesis that such algorithm may inevitably converge to trivial policy in the end.

Similar to what we have observed during discussion with @emailweixu, when such "dip" happens, the value network's dynamics still look good, but the training dynamics of the actor network is almost static:

1. All gradients = 0
2. The encoder's final ReLU always output 0

**Comment by breakds** [comment]
Btw, a "somewhat good" policy (on step 3000) looks like this:


https://user-images.githubusercontent.com/1111035/132715203-934d58ff-29a5-496c-bf6c-3aeaee5d24eb.mp4

**Comment by emailweixu** [comment]
> Yesterday after adding advantage normalization and replacing huber with MSE, such problem is almost solved until at the end it still converges to the trivial policy. The below graph compares the before (orange) and after (red) training curve. I think it is much alleviated but it seems that the algorithm cannot help converging to trivial policy at the end.

Perhaps you can try to train OpenAI's implementation for longer time to see if it also has this problem.

**Comment by breakds** [comment]
> > Yesterday after adding advantage normalization and replacing huber with MSE, such problem is almost solved until at the end it still converges to the trivial policy. The below graph compares the before (orange) and after (red) training curve. I think it is much alleviated but it seems that the algorithm cannot help converging to trivial policy at the end.
> 
> Perhaps you can try to train OpenAI's implementation for longer time to see if it also has this problem.

Sure, I will train it over the weekend with extended number of samples.

**Comment by breakds** [comment]
I took one round of training with the PPG implementation (with Auxiliary turned off, so this is equivalently PPO). The result looks good with batch size = 2048 and environment parallel = 64

![ppg_return](https://user-images.githubusercontent.com/1111035/132953808-4decf42d-e4c2-4426-aa81-0caf7bbbcecb.png)

On the other hand, the successful PPO training with batch size = 4096 and environment parallel = 96 looks like

![ppo](https://user-images.githubusercontent.com/1111035/132953826-e76d7b9a-1edb-4992-a86e-2fa84d393964.png)

Though this is not apple to apple, I think the PPG version PPO does meet the expectation.

**Comment by breakds** [comment]
When auxiliary is turned on, the memory consumption continues going up. I am working on fixing that.

**Comment by breakds** [comment]
There are two issues that prevents us from training Procgen + PPG/PPO on 3080, 10 GB Memory (which I think is a reasonable goal since OpenAI achieves that, and 3090 or higher end card are too expensive).

1. PPG here requires a full replay buffer to be recorded, where each step's observation (64 environment + 2048 mini batch size) took 3MB, and each iteration with 256 steps will take around 800MB of GPU memory if stored in GPU. This definitely needs to be stored in CPU instead, I am thinking about creating a new type of experience replayer to minimize the impact on algorithm.py
2. Even without auxiliary phase (degenerate to PPO), 64 environments + 2048 mini batch exceeds 10GB GPU memory significantly. However the observations only accounts for 800MB here so the 10GB memory footprint is not understandable to me. I am currently trying to figure this out.

**Comment by emailweixu** [comment]
> 2. Even without auxiliary phase (degenerate to PPO), 64 environments + 2048 mini batch exceeds 10GB GPU memory significantly. However the observations only accounts for 800MB here so the 10GB memory footprint is not understandable to me. I am currently trying to figure this out.

The intermediate results of the model take much more memory than its input.

**Comment by breakds** [comment]
For No.2, I tried to come back to the essential part and see the memory consumption. The following script keeps a replay buffer of 64 env x 256 step buffer of time steps. It took 943 MB. This simulates the replay buffer needed for our PPO. 

```python
import alf
import torch
import gym3
import time
from procgen import ProcgenGym3Env
from alf.examples import procgen_conf


def run_wrapped(num_envs=64, unroll_length=256):
    alf.config(
        'create_environment', env_name='bossfight', num_parallel_environments=num_envs)
    env = alf.get_env()
    step = 0
    replay_buffer = []
    while True:
        actions = env.action_spec().ones(outer_dims=(num_envs,))
        time_step = env.step(actions)

        replay_buffer.append(time_step)
        step += 1

        if step % unroll_length == 0:
            print(f'step: {step}, replay_buffer size: {len(replay_buffer)}')
            replay_buffer = []
            time.sleep(1)


if __name__ == '__main__':
    if torch.cuda.is_available():
        alf.set_default_device('cuda')
    run_wrapped()

```

while if `unroll_length=8192` (simulating No.1, 32 iterations), the memory consumption is 8879MB. 

The numbers are reasonable here because this is untransformed data so that observation is of `uint8` instead of `float32`.

**Comment by breakds** [comment]
> > 1. Even without auxiliary phase (degenerate to PPO), 64 environments + 2048 mini batch exceeds 10GB GPU memory significantly. However the observations only accounts for 800MB here so the 10GB memory footprint is not understandable to me. I am currently trying to figure this out.
> 
> The intermediate results of the model take much more memory than its input.

Fair enough, as there are a few convolutional layers, though some of them are actually down sampling the image by half. I'll do some calculation just to convince myself. Here applying lean function to each of the downsampling stack reduces only a few hundreds megabytes of memory consumption - I thought it would remove the intermediate result stored in the resulting tensor, but I might have misunderstanding here.

**Comment by hnyu** [comment]
> 1. PPG here requires a full replay buffer to be recorded, where each step's observation (64 environment + 2048 mini batch size) took 3MB, and each iteration with 256 steps will take around 800MB of GPU memory if stored in GPU. This definitely needs to be stored in CPU instead, I am thinking about creating a new type of experience replayer to minimize the impact on algorithm.py

For replay buffer, why do you count the batch size for the memory? I think batch size is for training and a batch has to be in GPU. The correct way to compute replay buffer memory for PPO is: ``num_envs * unroll_length * observation_size``.

**Comment by hnyu** [comment]
> > 1. PPG here requires a full replay buffer to be recorded, where each step's observation (64 environment + 2048 mini batch size) took 3MB, and each iteration with 256 steps will take around 800MB of GPU memory if stored in GPU. This definitely needs to be stored in CPU instead, I am thinking about creating a new type of experience replayer to minimize the impact on algorithm.py
> 
> For replay buffer, why do you count the batch size for the memory? I think batch size is for training and a batch has to be in GPU. The correct way to compute replay buffer memory for PPO is: `num_envs * unroll_length * observation_size`.

For PPG, if a larger replay buffer (not one-time replay) is needed, then it's already using CPU. Only the one-time replay buffer used by PPO is in GPU.

**Comment by breakds** [comment]
> For replay buffer, why do you count the batch size for the memory? I think batch size is for training and a batch has to be in GPU. The correct way to compute replay buffer memory for PPO is: `num_envs * unroll_length * observation_size`.

Sorry I wasn't clear about this. There are two parts of constantly occupied memory: the replay buffer and the network forward/backward. `batch_size` does not contribute to the first part but does for the second part.

> For PPG, if a larger replay buffer (not one-time replay) is needed, then it's already using CPU. Only the one-time replay buffer used by PPO is in GPU.

PPG needs to go over the whole replay buffer when auxiliary phase is activated, and my original implementation is using one-time replay buffer for it. This part needs to be changed for sure.

**Comment by breakds** [comment]
The final PR is in #1030 under review.

**Comment by breakds** [comment]
PR is submitted. Closing this issue.
